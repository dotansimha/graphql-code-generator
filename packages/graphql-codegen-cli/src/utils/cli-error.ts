import { DetailedError } from '@graphql-codegen/plugin-helpers';
import { isBrowser, isNode } from './is-browser';

type CompositeError = Error | DetailedError;
type ListrError = Error & { errors: CompositeError[] };
export function isListrError(err: Error & { name?: unknown; errors?: unknown }): err is ListrError {
  return err.name === 'ListrError' && Array.isArray(err.errors) && err.errors.length > 0;
}

export function cliError(err: any, exitOnError = true) {
  let msg: string;

  if (err instanceof Error) {
    msg = err.message || err.toString();
  } else if (typeof err === 'string') {
    msg = err;
  } else {
    msg = JSON.stringify(err);
  }

  // eslint-disable-next-line no-console
  console.error(msg);

  if (exitOnError && isNode) {
    process.exit(1);
  } else if (exitOnError && isBrowser) {
    throw err;
  }
}
