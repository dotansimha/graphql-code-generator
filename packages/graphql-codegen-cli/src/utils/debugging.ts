import { getLogger } from './logger';

let queue: Array<{
  message: string;
  meta?: any[];
}> = [];

export function debugLog(message: string, ...meta: any[]) {
  if (!process.env.GQL_CODEGEN_NODEBUG && process.env.DEBUG !== undefined) {
    queue.push({
      message,
      meta
    });
  }
}

export function printLogs() {
  if (!process.env.GQL_CODEGEN_NODEBUG && process.env.DEBUG !== undefined) {
    queue.forEach(log => {
      getLogger().info(log.message, ...log.meta);
    });
    resetLogs();
  }
}

export function resetLogs() {
  queue = [];
}
