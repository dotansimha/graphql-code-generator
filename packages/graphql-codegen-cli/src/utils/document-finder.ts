import * as stripComments from 'strip-comments';
import { parse } from 'graphql-codegen-core';

export const extractDocumentStringFromCodeFile = (fileContent: string): string | null => {
  try {
    const parsed = parse(fileContent);

    if (parsed) {
      return fileContent;
    }
  } catch (e) {
    try {
      fileContent = stripComments(fileContent, { sourceType: 'module' });
    } catch (e) {
      try {
        fileContent = stripComments(fileContent);
      } catch (e) {
        // nothing to to here
      }
    }

    let matches = fileContent.match(/gql[(]?`([\s\S\n\r.]*?)`/gm);

    if (matches === null) {
      matches = fileContent.match(/(['"](query|subscription|fragment|mutation) .*?['"])/gm);
    }

    const result = (matches || []).map(item => item.replace(/\$\{.*?\}/g, '').replace(/(gql|[(]?`)/g, '')).join();

    if (!result || result === '') {
      return null;
    } else {
      return result;
    }
  }
};
