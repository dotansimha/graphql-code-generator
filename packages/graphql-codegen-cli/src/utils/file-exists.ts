import fs from 'fs';

export function fileExists(filePath: string): boolean {
  try {
    return fs.statSync(filePath).isFile();
  } catch (err) {
    return false;
  }
}
