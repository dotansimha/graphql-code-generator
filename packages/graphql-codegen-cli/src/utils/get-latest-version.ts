import { fetch } from '@whatwg-node/fetch';

/**
 * Fetches the version directly from the registry instead of depending on
 * an ESM only module as latest-version does.
 * @param packageName
 */
export async function getLatestVersion(packageName: string): Promise<string> {
  return fetch(`https://unpkg.com/${packageName}/package.json`)
    .then(res => res.json())
    .then(pkg => pkg.version);
}
