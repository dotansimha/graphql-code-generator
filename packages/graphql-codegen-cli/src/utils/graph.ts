import { Types, normalizeInstanceOrArray } from '@graphql-codegen/plugin-helpers';
import { DepGraph } from 'dependency-graph';
import { resolve } from 'path';
import minimatch from 'minimatch';
import { isString, isUrl, isGlob } from './helpers';

// 1. Generate DepGraph
// 2. Sort `generates` so they won't stuck waiting for each other (on max concurrency limit hit)
// 3. Shared Promise per each output (so we know when output is generated)

type Graph = {
  config: Types.ConfiguredOutput;
  success: () => void;
  fail: (error: any) => void;
  running: Promise<void>;
};

/**
 * This way we wont't get unhandled rejection issues.
 * Jest is super sensible, it fails the test when unhandled promise is detected.
 */
function createLazyPromise() {
  let error: any;
  let resolved = false;
  let promise: Promise<void>;

  let resolve: () => void;
  let reject: (error: any) => void;

  return {
    resolve() {
      if (!resolved) {
        // postpone resolve()
        resolved = true;
      }

      if (resolve) {
        // if promise was created, call resolve()
        resolve();
      }
    },
    reject(reason: any) {
      if (!error) {
        // postpone reject()
        error = reason;
      }

      if (reject) {
        // if promise was created, call reject()
        reject(reason);
      }
    },
    get promise() {
      // we only create if there was an attempt to consume it

      // if promise was not created
      if (!promise) {
        if (error) {
          // and we already postponed reject()
          // reject() now
          promise = Promise.reject(error);
        } else if (resolved) {
          // and we already postponed resolve()
          // resolve() now
          promise = Promise.resolve();
        } else {
          // if there was no reject() or resolve() postponed
          // just create a new promise
          // eslint-disable-next-line promise/param-names
          promise = new Promise<void>((yes, not) => {
            resolve = yes;
            reject = not;
          });
        }
      }

      // return our promise
      return promise;
    },
  };
}

// Generate graph of dependencies
export function createGraph({ generates, cwd }: { generates: { [filename: string]: Types.ConfiguredOutput }; cwd: string }) {
  const graph = new DepGraph<Graph>({
    circular: false,
  });

  for (const output in generates) {
    if (generates.hasOwnProperty(output)) {
      const config = generates[output];
      const state = createLazyPromise();

      graph.addNode(output, {
        config,
        success: state.resolve,
        fail: state.reject,
        // we need it to avoid unhandled promise rejection (especially in jest)
        get running() {
          return state.promise;
        },
      });
    }
  }

  // pass only file paths and glob patterns
  const isStringOrGlob = (val: any): val is string => isString(val) && !isUrl(val);

  // check if source file matches the pointer (file path or glob)
  const matchFile = (src: string, pointer: string): boolean => {
    if (isGlob(pointer)) {
      return minimatch(src, pointer);
    }

    return resolve(cwd, pointer) === resolve(cwd, src);
  };

  // set dependencies
  function findDependencies(src: string) {
    Object.keys(generates)
      // skip our output
      .filter(key => key !== src)
      // iterate over the rest
      .forEach(output => {
        const { config } = graph.getNodeData(output);
        // normalize so we get arrays (always)
        const outputSpecificSchemas: string[] = normalizeInstanceOrArray<Types.Schema>(config.schema).filter(isStringOrGlob);
        const outputSpecificDocuments = normalizeInstanceOrArray<Types.OperationDocument>(config.documents).filter(isStringOrGlob);

        outputSpecificSchemas.forEach(pointer => {
          if (matchFile(src, pointer)) {
            // our source depends on this output
            graph.addDependency(output, src);
          }
        });

        outputSpecificDocuments.forEach(pointer => {
          if (matchFile(src, pointer)) {
            // our source depends on this output
            graph.addDependency(output, src);
          }
        });
      });
  }

  // Look for dependencies in each output
  for (const output in generates) {
    if (generates.hasOwnProperty(output)) {
      findDependencies(output);
    }
  }

  return graph;
}

export async function waitForDependencies({ graph, output }: { graph: DepGraph<Graph>; output: string }): Promise<void> {
  // wait for other outputs to resolve, the one that we depend on
  await Promise.all(
    graph
      .dependenciesOf(output)
      .filter(Boolean)
      .map(dep => graph.getNodeData(dep).running)
  );
}
