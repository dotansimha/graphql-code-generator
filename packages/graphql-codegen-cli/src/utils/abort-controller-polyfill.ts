import { EventEmitter } from 'events';
import { debugLog } from './debugging.js';

/**
 * Node v14 does not have AbortSignal or AbortController, so to safely use it in
 * another module, you can import it from here.
 *
 * Node v14.7+ does have it, but only with flag --experimental-abortcontroller
 *
 * We don't actually use AbortController anywhere except in tests, but it
 * still gets called in watcher.ts, so by polyfilling it we can avoid breaking
 * existing installations using Node v14 without flag --experimental-abortcontroller,
 * and we also ensure that tests continue to pass under Node v14 without any new flags.
 *
 * This polyfill was adapted (TypeScript-ified) from here:
 * https://github.com/southpolesteve/node-abort-controller/blob/master/index.js
 */

class AbortSignalPolyfill implements AbortSignal {
  eventEmitter: EventEmitter;
  onabort: EventListener;
  aborted: boolean;
  reason: any | undefined;

  constructor() {
    this.eventEmitter = new EventEmitter();
    this.onabort = null;
    this.aborted = false;
    this.reason = undefined;
  }
  toString() {
    return '[object AbortSignal]';
  }
  get [Symbol.toStringTag]() {
    return 'AbortSignal';
  }
  removeEventListener(name, handler) {
    this.eventEmitter.removeListener(name, handler);
  }
  addEventListener(name, handler) {
    this.eventEmitter.on(name, handler);
  }
  // @ts-expect-error No Event type in Node 14
  dispatchEvent(type: string) {
    const event = { type, target: this };
    const handlerName = `on${event.type}`;

    if (typeof this[handlerName] === 'function') this[handlerName](event);

    return this.eventEmitter.emit(event.type, event);
  }
  throwIfAborted() {
    if (this.aborted) {
      throw this.reason;
    }
  }
  static abort(reason: any) {
    const controller = new AbortController();
    controller.abort(reason);
    return controller.signal;
  }
  static timeout(time) {
    const controller = new AbortController();
    setTimeout(() => controller.abort(new Error('TimeoutError')), time);
    return controller.signal;
  }
}
const AbortSignal = global.AbortSignal ?? AbortSignalPolyfill;

class AbortControllerPolyfill implements AbortController {
  signal: AbortSignal;

  constructor() {
    debugLog('Using polyfilled AbortController');
    // @ts-expect-error No Event type in Node 14
    this.signal = new AbortSignal();
  }
  abort(reason?: any) {
    if (this.signal.aborted) return;

    // @ts-expect-error Not a read only property when polyfilling
    this.signal.aborted = true;

    if (reason) {
      // @ts-expect-error Not a read only property when polyfilling
      this.signal.reason = reason;
    } else {
      // @ts-expect-error Not a read only property when polyfilling
      this.signal.reason = new Error('AbortError');
    }

    // @ts-expect-error No Event type in Node 14
    this.signal.dispatchEvent('abort');
  }
  toString() {
    return '[object AbortController]';
  }
  get [Symbol.toStringTag]() {
    return 'AbortController';
  }
}

const AbortController = global.AbortController ?? AbortControllerPolyfill;

export { AbortController };
