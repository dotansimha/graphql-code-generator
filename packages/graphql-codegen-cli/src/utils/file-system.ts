import * as fs from 'fs';
import { fileExists } from './file-exists';

export default {
  writeSync(filepath: string, content: string) {
    fs.writeFileSync(filepath, content);
  },
  fileExists
};
