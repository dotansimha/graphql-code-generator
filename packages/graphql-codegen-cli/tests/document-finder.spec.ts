import * as fs from 'fs';
import { parse } from 'graphql';
import { Source } from 'graphql-codegen-core';

import { extractDocumentStringFromCodeFile } from '../src/utils/document-finder';

describe('extractDocumentStringFromCodeFile', () => {
  function extract(fileName: string) {
    const fileContent = fs.readFileSync(`./tests/test-files/${fileName}`).toString();
    const doc = extractDocumentStringFromCodeFile(fileContent);
    expect(tryParse(doc)).not.toThrow();
    return doc.trim();
  }

  function tryParse(doc: string) {
    return () => parse(new Source(doc));
  }

  it('file with gql and template literal', () => {
    const doc = extract('1.ts');
    expect(doc).toMatchSnapshot();
  });

  it('file with gql and string and function use', () => {
    const doc = extract('2.ts');
    expect(doc).toMatchSnapshot();
  });

  it('file with simple gql content', () => {
    const doc = extract('3.graphql');
    expect(doc).toMatchSnapshot();
  });

  it('file with commented code', () => {
    const doc = extract('4.ts');
    expect(doc).toMatchSnapshot();
  });

  it('file with gql and template literal in ts/tsx file', () => {
    const doc = extract('5.tsx');
    expect(doc).toMatchSnapshot();
  });

  it('file with gql and template literal, with use of string variables', () => {
    const doc = extract('6.ts');
    expect(doc).toMatchSnapshot();
  });
});
