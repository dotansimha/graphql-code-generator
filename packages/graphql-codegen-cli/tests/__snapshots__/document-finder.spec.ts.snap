// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractDocumentStringFromCodeFile file with commented code 1`] = `
"query myQuery2 {
    data {
      field1
      field2
    }
  }"
`;

exports[`extractDocumentStringFromCodeFile file with gql and string and function use 1`] = `
"query myQuery {
    data {
      field1
      field2
    }
  }"
`;

exports[`extractDocumentStringFromCodeFile file with gql and template literal 1`] = `
"query myQuery {
    data {
      field1
      field2
    }
  }"
`;

exports[`extractDocumentStringFromCodeFile file with gql and template literal in ts/tsx file 1`] = `
"query GPlayerList {
    allPlayers {
      id
      tier
      name
    }
  }"
`;

exports[`extractDocumentStringFromCodeFile file with gql and template literal, with use of string variables 1`] = `
"query ArticleDetail {
    Article(id: \\"thisisnotarealID\\") {
      title
    }
  }"
`;

exports[`extractDocumentStringFromCodeFile file with simple gql content 1`] = `
"query myQuery {
  data {
    field1
    field2
  }
}"
`;
